keps/420-partial-admission/README.md

```diff
@@ -0,0 +1,139 @@
+# KEP-420: Allow partial admission of PodSets
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Workload API](#workload-api)
+  - [Scheduler / Flavorassignment](#scheduler--flavorassignment)
+  - [Jobframework](#jobframework)
+  - [batch/Job controller](#batchjob-controller)
+  - [Test Plan](#test-plan)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+Add and optional way of allowing the partial admission of a workload if the full admission is not possible.
+
+## Motivation
+
+In practice, not all Workloads require the parallel execution of all the `count` of a `PodSet`, for such cases having a way to partially reserve the quota in order to prevent starvation.
+
+For example if a batch/Job has parallelism is x and there is only quota available for y < x then jobs should still be admitted.
+
+### Goals
+
+Provide an opt-in way for Workloads to accept the admission with a lower count of pods if the full count is not available.
+
+### Non-Goals
+
+Since this is an opt-in feature, the parent job should accept the partial admission parameters provided by Kueue.
+
+Kueue will not take any measure to ensure that the parent job respects the assigned quota.
+
+## Proposal
+
+Change the way the flavor assigner work to support decrementing the pods count in order to find a better fit for the current workload.
+In case a partial fit is chosen, the jobframework reconciler should provide the admitted pod counts to the parent job before unsuspending it, in a similar fashion as the node selectors are provided. In case the job gets suspended, the original pod counts should be restored in order to allow a potential future admission with its original pod counts.
+
+### User Stories
+
+Kueue issue [420](https://github.com/kubernetes-sigs/kueue/issues/420) provides details on the initial feature details and its applicability for `batch/Job`.
+
+### Risks and Mitigations
+
+TBD: It should be decided if the partial admission should be taken into account before attempting preemption. Since preemption has its own penalties (time to evict the active workloads, potential waste of the resources used by the preempted workloads) I vote to take it into account before.
```

Comment:
yes

Reply:
I think we should try to minimize preemption.

But there are two forms of preemption: within ClusterQueue and within cohort.

### within Cluster Queue

I agree with the chosen option for 1.A.a
But we should do 1.A.b.2: accept a partial admission with preemption

### within cohort

In this case, the ClusterQueue could be reclaiming quota that belongs to it by doing preemption before going into partial admission.

Can we do this distinction? The implementation might be a bit more complex.

cc @ahg-g and @mwielgus for comment on the correct semantics.

Reply:
Indeed the implementation could be complex. Also it might not be very easy to follow.

Reply:
I was discussing this offline with @mwielgus and he convinced me that preemption should have higher priority.

In the case of cohort, the ClusterQueue is reclaiming quota that belongs to it, so it certainly should preempt.
In the case of higher priority within the ClusterQueue, the Workload has higher priority, so it should preempt. Otherwise it would depend on the order in which workloads come in:
- if low priority comes first, it's running and the higher priority workload has less pods.
- if high priority comes first, then it has a big number of pods and low priority doesn't run.

Prioritizing preemption has the effect of behaving similar to the case where the high priority workload comes first.

Reply:
I like the opinions of @mwielgus. The suggestion sounds consistent for preemption behavior.

Reply:
ok